File: arena-container/src/identifier.rs
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::OnceLock;
use std::time::{SystemTime, UNIX_EPOCH};

const SUFFIX_LEN: usize = 6;
const MODULE_PREFIX: &str = "arena-";
const BASE36: &[u8; 36] = b"0123456789abcdefghijklmnopqrstuvwxyz";

pub fn build(module: &str, name: &str) -> String {
    if is_already_built(name) {
        return name.to_string();
    }
    let slug = slugify(name);
    let suffix = new_suffix();
    if slug.is_empty() {
        format!("{module}-{suffix}")
    } else {
        format!("{module}-{slug}-{suffix}")
    }
}

pub fn sanitize_for_container(identifier: &str) -> String {
    slugify(identifier)
}

pub fn resolve_container_name(identifier: &str, override_name: Option<&str>) -> String {
    match override_name {
        Some(name) => name.to_string(),
        None => sanitize_for_container(identifier),
    }
}

fn slugify(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    let mut last_dash = false;
    for c in s.chars() {
        let c = c.to_ascii_lowercase();
        if c.is_ascii_alphanumeric() {
            out.push(c);
            last_dash = false;
        } else if !last_dash {
            out.push('-');
            last_dash = true;
        }
    }
    out.trim_matches('-').to_string()
}

fn new_suffix() -> String {
    static SEED: OnceLock<u64> = OnceLock::new();
    static COUNTER: AtomicU64 = AtomicU64::new(0);

    let seed = *SEED.get_or_init(|| {
        let nanos = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map(|d| d.as_nanos() as u64)
            .unwrap_or(0);
        nanos ^ (std::process::id() as u64).rotate_left(32)
    });
    let n = seed.wrapping_add(COUNTER.fetch_add(1, Ordering::Relaxed));
    to_base36(n)
}

fn to_base36(mut n: u64) -> String {
    let mut buf = [b'0'; SUFFIX_LEN];
    for slot in buf.iter_mut().rev() {
        *slot = BASE36[(n % 36) as usize];
        n /= 36;
    }
    String::from_utf8(buf.to_vec()).expect("base36 digits are ascii")
}

fn is_already_built(name: &str) -> bool {
    let Some(rest) = name.strip_prefix(MODULE_PREFIX) else {
        return false;
    };
    let Some(suffix) = rest.rsplit('-').next() else {
        return false;
    };
    suffix.len() == SUFFIX_LEN
        && suffix
            .chars()
            .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit())
}

File: arena-pytest/src/arena_pytest/support/_identifier.py
from __future__ import annotations

import itertools
import os
import time

_SUFFIX_LEN = 6
_ALPHABET = "0123456789abcdefghijklmnopqrstuvwxyz"
_BASE = len(_ALPHABET)
_MASK_64 = 0xFFFFFFFFFFFFFFFF


def _seed() -> int:
    if not hasattr(_seed, "_value"):
        nanos = time.time_ns() & _MASK_64
        pid = os.getpid() & _MASK_64
        pid_rot = ((pid << 32) | (pid >> 32)) & _MASK_64
        _seed._value = nanos ^ pid_rot
    return _seed._value


_COUNTER = itertools.count()


def _slugify(s: str) -> str:
    out = []
    last_dash = False
    for c in s:
        c = c.lower()
        if c.isascii() and c.isalnum():
            out.append(c)
            last_dash = False
        elif not last_dash:
            out.append("-")
            last_dash = True
    return "".join(out).strip("-")


def _new_suffix() -> str:
    n = (_seed() + next(_COUNTER)) & _MASK_64
    digits = []
    for _ in range(_SUFFIX_LEN):
        digits.append(_ALPHABET[n % _BASE])
        n //= _BASE
    return "".join(reversed(digits))


def _has_suffix(name: str) -> bool:
    if "--" not in name:
        return False
    suffix = name.rsplit("--", 1)[-1]
    if len(suffix) != _SUFFIX_LEN:
        return False
    return all(c in _ALPHABET for c in suffix)


def build(module: str, name: str) -> str:
    if _has_suffix(name):
        return name
    slug = _slugify(name)
    suffix = _new_suffix()
    if not slug:
        return f"{module}-{suffix}"
    return f"{module}-{slug}--{suffix}"
